group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java

```diff
@@ -939,12 +941,20 @@ public static ConsumerGroup fromClassicGroup(
         consumerGroup.setTargetAssignmentEpoch(classicGroup.generationId());
 
         classicGroup.allMembers().forEach(classicGroupMember -> {
-            Map<Uuid, Set<Integer>> assignedPartitions = toTopicPartitionMap(
-                ConsumerProtocol.deserializeConsumerProtocolAssignment(
-                    ByteBuffer.wrap(classicGroupMember.assignment())
-                ),
-                topicsImage
-            );
+            // The assigned partition can be empty if the member just joined and has never synced.
+            // We should accept the empty assignment.
+            Map<Uuid, Set<Integer>> assignedPartitions;
+            if (Arrays.equals(classicGroupMember.assignment(), EMPTY_ASSIGNMENT)) {
+                assignedPartitions = Collections.emptyMap();
+            } else {
+                assignedPartitions = toTopicPartitionMap(
+                    ConsumerProtocol.deserializeConsumerProtocolAssignment(
+                        ByteBuffer.wrap(classicGroupMember.assignment())
+                    ),
+                    topicsImage
+                );
+            }
+
             ConsumerProtocolSubscription subscription = ConsumerProtocol.deserializeConsumerProtocolSubscription(
                 ByteBuffer.wrap(classicGroupMember.metadata(classicGroup.protocolName().get()))
```

Comment:
For my understanding. I think that we don't need a similar fix here because every member is guaranteed to have metadata set when it joins. However, the metadata could be malformed. In this case, the deserialization fails. I suppose that it results in a unknown server error returned to the client. Is my understanding correct?

Reply:
> For my understanding. I think that we don't need a similar fix here because every member is guaranteed to have metadata set when it joins. However, the metadata could be malformed. In this case, the deserialization fails.

This is correct.

> it results in a unknown server error returned to the client

In such case, a `Group_Id_Not_found` is returned to follow the upgrade validation failure returning `Group_Id_Not_found` in `getOrMaybeCreateConsumerGroup`.
https://github.com/apache/kafka/blob/615c8c0e11678c39904d8bdde991a0c1a4b67625/group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java#L1204-L1211